README.md
# MMM-Thingiverse

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

Display 3D printable objects found on Thingiverse
![Screenshot from 2021-02-20 07-37-07](https://user-images.githubusercontent.com/22528729/108597418-89b64800-734e-11eb-90c5-a78c7f01bd71.png)
![Screenshot from 2021-02-20 07-37-33](https://user-images.githubusercontent.com/22528729/108597419-8a4ede80-734e-11eb-8643-2063ada05341.png)
![Screenshot from 2021-02-20 07-38-44](https://user-images.githubusercontent.com/22528729/108597457-c41fe500-734e-11eb-8f7f-ff8b95c21a26.png)
![Screenshot from 2021-02-20 07-40-01](https://user-images.githubusercontent.com/22528729/108597474-ddc12c80-734e-11eb-92ec-2fbdcc4ba0ab.png)
![Screenshot from 2021-02-20 07-40-07](https://user-images.githubusercontent.com/22528729/108597475-de59c300-734e-11eb-82c3-f52ef78004f5.png)

## Quickstart

```bash
cd ~/MagicMirror/modules
git clone https://github.com/appdevelopmentandsuch/MMM-Thingiverse
```

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:

```js
var config = {
  modules: [
    {
      module: 'MMMM-Thingiverse',
      config: {
        appToken: '',
      },
    },
  ],
};
```

## Configuration options

| Option               | Description                                                                                                                                                                                                                                                                                     |
| -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `appToken`           | _Required_ A token aquired from the Thingiverse REST API to fetch things. You can aquire a token by going to https://www.thingiverse.com/apps/create and creating a **Web App**                                                                                                                 |
| `updateInterval`     | _Optional_ The time in milliseconds before switching to the next thing. <br><br>**Type:** `int`(milliseconds) <br>Default 60000 milliseconds (1 minute)                                                                                                                                         |
| `retryDelay`         | _Optional_ The time in milliseconds before retrying the Thingiverse REST API due to a previous failure to get things. <br><br>**Type:** `int`(milliseconds) <br>Default 5000 milliseconds (5 seconds)                                                                                           |
| `thingCount`         | _Optional_ The number of things you wish to grab and cycle through. <br><br>**Type:** `int`(count) <br>Default 100                                                                                                                                                                              |
| `startAtRandom`      | _Optional_ Start at a random position in the things. <br><br>**Type:** `boolean`(true/false) <br>Default false                                                                                                                                                                                  |
| `numThingsDisplayed` | _Optional_ Display 1, 3, or 5 things at any given time. <br><br>**Type:** `int`(1,3,5) <br>Default 1                                                                                                                                                                                            |
| `category`           | _Optional_ Display things based on a provided category. A list of valid categories can be grabbed from the Thingiverse API using the endpoint `https://api.thingiverse.com/categories?access_token=[YOUR_APP_TOKEN]`. Use the `slug` entry for this prop <br><br>**Type:** `string` <br>Default |


CHANGELOG.md
# MagicMirror-Modules-Thingiverse Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.1] - Released

### Bug Fixes

- Fixed issue where the `page` variable would be set to `0` and sent in the API request, resulting in no data.
- Added validation ensuring that `things` are only set to an array, not data.

## [1.1.0] - Released

### Features

- Added a `category` property, allowing users to display things based on a specific category.
- Added a `isFeatured` property, allowing users to display things that were / are featured on Thingiverse.
- Added a `numThingsDisplayed` property, allowing users to display 1, 3, or 5 things at any given time.
- Added a `searchBy` property, allowing users to display either newest things, or popular things.
- Added safety checking and fallback, such that if there is an error pulling things for a category, it will revert to displaying popular / newest things, whatever your config is set to.
- Added alt image text in the event the thumbnail image could not be displayed.

### Bug Fixes

- Removed the QR Code display from the module, QR Code generation was not properly working.
- Fixed issue where, when all things had been displayed and the next things were getting ready, the screen would be blank.

## [1.0.0] - Released

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MagicMirror-Modules-Thingiverse
 *
 * By 
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");

module.exports = NodeHelper.create({

	// Override socketNotificationReceived method.

	/* socketNotificationReceived(notification, payload)
	 * This method is called when a socket notification arrives.
	 *
	 * argument notification string - The identifier of the noitication.
	 * argument payload mixed - The payload of the notification.
	 */
	socketNotificationReceived: function(notification, payload) {
		if (notification === "MagicMirror-Modules-Thingiverse-NOTIFICATION_TEST") {
			console.log("Working notification system. Notification:", notification, "payload: ", payload);
			// Send notification
			this.sendNotificationTest(this.anotherFunction()); //Is possible send objects :)
		}
	},

	// Example function send notification test
	sendNotificationTest: function(payload) {
		this.sendSocketNotification("MagicMirror-Modules-Thingiverse-NOTIFICATION_TEST", payload);
	},

	// this you can create extra routes for your module
	extraRoutes: function() {
		var self = this;
		this.expressApp.get("/MagicMirror-Modules-Thingiverse/extra_route", function(req, res) {
			// call another function
			values = self.anotherFunction();
			res.send(values);
		});
	},

	// Test another function
	anotherFunction: function() {
		return {date: new Date()};
	}
});


MagicMirror-Modules-Thingiverse.js
/* global Module */

/* Magic Mirror
 * Module: MagicMirror-Modules-Thingiverse
 *
 * By
 * MIT Licensed.
 */

Module.register('MagicMirror-Modules-Thingiverse', {
  defaults: {
    appToken: '',
    updateInterval: 60000,
    retryDelay: 5000,
  },

  requiresVersion: '2.1.0',

  start: function () {
    var self = this;
    var dataRequest = null;
    var dataNotification = null;

    this.loaded = false;

    this.getData();
    setInterval(function () {
      self.updateDom();
    }, this.config.updateInterval);
  },

  getData: function () {
    var self = this;

    var urlApi = `https://api.thingiverse.com/search/?sort=popular&access_token=${appToken}`;
    var retry = true;

    var dataRequest = new XMLHttpRequest();
    dataRequest.open('GET', urlApi, true);
    dataRequest.onreadystatechange = function () {
      console.log(this.readyState);
      if (this.readyState === 4) {
        console.log(this.status);
        if (this.status === 200) {
          self.processData(JSON.parse(this.response));
        } else if (this.status === 401) {
          self.updateDom(self.config.animationSpeed);
          Log.error(self.name, this.status);
          retry = false;
        } else {
          Log.error(self.name, 'Could not load data.');
        }
        if (retry) {
          self.scheduleUpdate(self.loaded ? -1 : self.config.retryDelay);
        }
      }
    };
    dataRequest.send();
  },

  scheduleUpdate: function (delay) {
    // var nextLoad = this.config.updateInterval;
    // if (typeof delay !== 'undefined' && delay >= 0) {
    //   nextLoad = delay;
    // }
    // nextLoad = nextLoad;
    // var self = this;
    // setTimeout(function () {
    //   self.getData();
    // }, nextLoad);
  },

  getDom: function () {
    var self = this;

    var wrapper = document.createElement('div');
    if (this.dataRequest) {
      var wrapperDataRequest = document.createElement('div');
      wrapperDataRequest.innerHTML = this.dataRequest.title;

      var labelDataRequest = document.createElement('label');
      labelDataRequest.innerHTML = this.translate('TITLE');

      wrapper.appendChild(labelDataRequest);
      wrapper.appendChild(wrapperDataRequest);
    }

    if (this.dataNotification) {
      var wrapperDataNotification = document.createElement('div');

      wrapperDataNotification.innerHTML =
        this.translate('UPDATE') + ': ' + this.dataNotification.date;

      wrapper.appendChild(wrapperDataNotification);
    }
    return wrapper;
  },

  getScripts: function () {
    return [];
  },

  getStyles: function () {
    return ['MagicMirror-Modules-Thingiverse.css'];
  },

  getTranslations: function () {
    return {
      en: 'translations/en.json',
    };
  },

  processData: function (data) {
    var self = this;
    this.dataRequest = data;
    if (this.loaded === false) {
      self.updateDom(self.config.animationSpeed);
    }
    this.loaded = true;

    this.sendSocketNotification(
      'MagicMirror-Modules-Thingiverse-NOTIFICATION_TEST',
      data,
    );
  },

  socketNotificationReceived: function (notification, payload) {
    if (notification === 'MagicMirror-Modules-Thingiverse-NOTIFICATION_TEST') {
      this.dataNotification = payload;
      this.updateDom();
    }
  },
});


MMM-Thingiverse.js
/* global Module */

/* Magic Mirror
 * Module: MMM-Thingiverse
 *
 * By Lucas Bock
 * MIT Licensed.
 */

Module.register('MMM-Thingiverse', {
  defaults: {
    appToken: '',
    retryDelay: 5000,
    startAtRandom: false,
    thingCount: 100,
    updateInterval: 60000,
    numThingsDisplayed: 1,
    searchBy: 'popular',
    isFeatured: false,
    category: '',
  },

  requiresVersion: '2.1.0',

  start: function () {
    var self = this;
    var dataRequest = null;

    this.currentPage = 1;
    this.currentThingId = -1;
    this.iterations = 0;
    this.loaded = false;
    this.maxPages = 11;
    this.newRequestTimeout = 10000;
    this.overrideUrl = null;
    this.things = { hits: [] };

    if (!['popular', 'newest'].includes(self.config.searchBy)) {
      self.config.searchBy = 'popular';
    }

    if (![1, 3, 5].includes(self.config.numThingsDisplayed)) {
      self.config.numThingsDisplayed = 3;
    }

    this.getData();

    setInterval(function () {
      self.updateDom();
    }, this.config.updateInterval);
  },

  getData: function () {
    var self = this;

    var apiBase = 'https://api.thingiverse.com/';

    var apiConfig = `access_token=${this.config.appToken}&per_page=${this.config.thingCount}&page=${self.currentPage}`;

    var popularConfig = `search/?sort=${this.config.searchBy}&is_featured=${
      this.config.isFeatured ? 1 : 0
    }`;

    var categoryConfig = `categories/${this.config.category}/things`;
    var urlApi =
      apiBase +
      (this.config.category ? categoryConfig + '?' : popularConfig + '&') +
      apiConfig;

    var fallbackUrl = apiBase + popularConfig + '&' + apiConfig;

    var retry = true;

    if (self.overrideUrl != null && self.overrideUrl !== urlApi) {
      urlApi = self.overrideUrl;
    }

    var dataRequest = new XMLHttpRequest();
    dataRequest.open('GET', urlApi, true);
    dataRequest.onreadystatechange = function () {
      if (this.readyState === 4) {
        if (this.status === 200) {
          self.iterations = 0;
          var parsedResponse = JSON.parse(this.response);
          if (
            (parsedResponse && Array.isArray(parsedResponse)) ||
            (parsedResponse.hits && Array.isArray(parsedResponse.hits))
          ) {
            self.things =
              self.config.category && self.overrideUrl === null
                ? parsedResponse
                : parsedResponse.hits;
            self.currentThingId = self.config.startAtRandom
              ? Math.floor(Math.random() * self.things.length)
              : 0;
            self.currentPage = Math.max(
              (self.currentPage + 1) % self.maxPages,
              1,
            );
            self.processData(self.things);
          }
        } else if (this.status === 401) {
          self.updateDom(self.config.animationSpeed);
          Log.error(self.name, this.status);
          self.overrideUrl = fallbackUrl;
          self.scheduleUpdate(self.loaded ? -1 : self.config.retryDelay);
        } else {
          Log.error(self.name, 'Could not load data.');
        }
      }
    };
    dataRequest.send();
  },

  scheduleUpdate: function (delay) {
    var nextLoad = this.config.updateInterval;
    if (typeof delay !== 'undefined' && delay >= 0) {
      nextLoad = delay;
    }
    nextLoad = nextLoad;
    var self = this;

    setTimeout(function () {
      self.getData();
    }, nextLoad);
  },

  createThingElement: function (thing) {
    var thingCard = document.createElement('div');
    thingCard.classList.add('MMM-Thingiverse-card');

    if (thing) {
      var thingCreator = document.createElement('div');
      thingCreator.classList.add('MMM-Thingiverse-creator');
      thingCreator.innerHTML = `<i>${thing.creator.name}</i>`;

      var thingName = document.createElement('div');
      thingName.classList.add('MMM-Thingiverse-name');
      thingName.innerHTML = `<b>${thing.name}</b>`;

      var thingThumbnail = document.createElement('img');
      thingThumbnail.classList.add('MMM-Thingiverse-thumbnail');
      thingThumbnail.src = thing.preview_image;
      thingThumbnail.alt = thing.name;

      thingCard.appendChild(thingName);
      thingCard.appendChild(thingThumbnail);
      thingCard.appendChild(thingCreator);
    }

    return thingCard;
  },

  getDom: function () {
    var self = this;

    var wrapper = document.createElement('div');
    wrapper.classList.add('MMM-Thingiverse-wrapper');

    if (this.dataRequest) {
      for (var i = 0; i < self.config.numThingsDisplayed; i++) {
        var thing = self.things[self.currentThingId % self.things.length];
        if (thing) {
          self.iterations = self.iterations + 1;

          if (self.iterations >= self.things.length) {
            self.scheduleUpdate(self.newRequestTimeout);
          }

          self.currentThingId = self.currentThingId + 1;
          wrapper.appendChild(self.createThingElement(thing));
        }
      }
    }

    return wrapper;
  },

  getScripts: function () {
    return [];
  },

  getStyles: function () {
    return ['MMM-Thingiverse.css'];
  },

  getTranslations: function () {
    return {
      en: 'translations/en.json',
    };
  },

  processData: function (data) {
    var self = this;

    if (Array.isArray(data)) {
      this.dataRequest = data;
      if (this.loaded === false) {
        self.updateDom(self.config.animationSpeed);
      }
      this.loaded = true;
    }
  },

  socketNotificationReceived: function (notification, payload) {},
});
